### PR TITLE
sass/node-sass -> sass/dart-sass

### PR DESCRIPTION
**Pull request by Firfi**
To overcome the issue https://github.com/Meteor-Community-Packages/meteor-scss/issues/294 we can switch to https://www.npmjs.com/package/sass .

This code works quite well for my project. Our css person uses :not selector extensively so it's not viable to change codebase.

Also, the original node-sass issue (https://github.com/sass/node-sass/issues/2330) seems to exist for a couple years already, therefore it's not clear when and if it will be resolved.

### CODE REVIEW

**Comment by znerol** [comment]
I wholeheartedly agree with swapping out `node-sass` for `dart-sass`. The PR needs some love though, review follows.

File: plugin/compile-scss.js [hunk 1]
```diff
@@ -80,7 +82,7 @@ class SassCompiler extends MultiFileCachingCompiler {
       return result && {
         data: result.css,
         sourceMap: result.sourceMap,
-        };
+      };
```
Comment (znerol) [inline plugin/compile-scss.js:85]: Avoid changing coding style in unrelated places if possible (even if the change of whitespace is correct). This adds noise to the PR and makes review harder.

File: plugin/compile-scss.js [hunk 2]
```diff
@@ -105,13 +107,13 @@ class SassCompiler extends MultiFileCachingCompiler {
...
       //If the referenced file has no extension, try possible extensions, starting with extension of the parent file.
-      let possibleExtensions = ['scss','sass','css'];
+      let possibleExtensions = [...expectedExtensions];
 
       if(! importPath.match(/\.s?(a|c)ss$/)){
         possibleExtensions = [
           inputFile.getExtension(),
           ...possibleExtensions.filter(e => e !== inputFile.getExtension())
-          ]
+        ]
```
Comment (znerol) [inline plugin/compile-scss.js:116]: Same here. Avoid unnecessary changes.

File: plugin/compile-scss.js [hunk 3]
```diff
@@ -126,13 +128,28 @@ class SassCompiler extends MultiFileCachingCompiler {
...
+          possibleFiles.push('{}/' + possibleFile);
+        }
+      }
+
+      for (const possibleFile of possibleFiles) {
+        if (possibleFile.startsWith('./') && !possibleFile.startsWith('{}') && !isAbsolute) {
+          possibleFiles.push('{}/' + possibleFile.substring('./'.length));
+        }
+      }
+
```
Comment (znerol) [inline plugin/compile-scss.js:143]: This changes the `possibleFiles` array while looping over it. Avoid that and instead do something like this:

```javascript

if (!isAbsolute) {
  let additionalPossibleFiles = [];
  for (const possibleFile of possibleFiles) {
    if(...) {
      additionalPossibleFiles.push(...)
    }
  }
  possibleFiles = possibleFiles.concat(additionalPossibleFiles);
}
```

Also I'm wondering why it is necessary to add those variants. Is this something required by `dart-sass` or is it an additional feature you like to include in `meteor-scss` package?

File: plugin/compile-scss.js [hunk 4]
```diff
@@ -126,13 +128,28 @@ class SassCompiler extends MultiFileCachingCompiler {
...
       //Try if one of the possible files exists
       for (const possibleFile of possibleFiles) {
         if ((isAbsolute && fileExists(possibleFile)) || (!isAbsolute && allFiles.has(possibleFile))) {
-            return { absolute: isAbsolute, path: possibleFile };
+          return { absolute: isAbsolute, path: possibleFile };
```
Comment (znerol) [inline plugin/compile-scss.js:147]: Unrelated change.

File: plugin/compile-scss.js [hunk 5]
```diff
@@ -126,13 +128,28 @@ class SassCompiler extends MultiFileCachingCompiler {
...
         }
       }
 
+      console.debug('imported style not found in possible files', possibleFiles);
+
```
Comment (znerol) [inline plugin/compile-scss.js:152]: Stray debug message. Please remove.

File: plugin/compile-scss.js [hunk 6]
```diff
@@ -257,14 +274,23 @@ function _getRealImportPathFromIncludes(importPath, getRealImportPathFn){
...
 
-    if (foundFile) {
-      return foundFile;
+  for (let includePath of _includePaths) {
+    for (let possibleFile of possibleFiles) {
+      possibleFilePath = path.join(includePath, possibleFile);
+      foundFile = getRealImportPathFn(possibleFilePath);
+      if (foundFile) {
+        return foundFile;
+      }
```
Comment (znerol) [inline plugin/compile-scss.js:293]: Same question here: Is this something required by `dart-sass` or is it an additional feature you like to include in `meteor-scss` package?

File: plugin/compile-scss.js [hunk 7]
```diff
@@ -297,6 +323,7 @@ function _loadIncludePaths(config) {
   } else {
     _includePaths = [];
   }
+
```
Comment (znerol) [inline plugin/compile-scss.js:326]: Unrelated.

File: plugin/compile-scss.js [hunk 8]
```diff
@@ -305,7 +332,8 @@ function _loadIncludePaths(config) {
  * @private
  */
 function _loadConfigurationFile() {
-  return _getConfig('scss-config.json') || {};
+  return _getConfig('scss-config.json') || {
+  };
```
Comment (znerol) [inline plugin/compile-scss.js:336]: Unrelated.

**Review by znerol — CHANGES_REQUESTED** [review]
Thanks for your work so far. The PR looks promising.

File: plugin/compile-scss.js (hunk 3 above)
Comment (Firfi) [reply plugin/compile-scss.js:143]: > This changes the possibleFiles
I'd gladly do it but I tried to go along the project code style. It's already done it a bunch of places. Should we go forward with immutability in only one place, leaving it mutable in older code? If not, shouldn't it be a matter of a separate pull request (probably, together with code formatting)?

File: plugin/compile-scss.js (hunk 1 above)
Comment (Firfi) [reply plugin/compile-scss.js:85]: I'm moving it to a separate PR

File: plugin/compile-scss.js (hunk 2 above)
Comment (Firfi) [reply plugin/compile-scss.js:116]: I'm moving it to a separate PR

File: plugin/compile-scss.js (hunk 4 above)
Comment (Firfi) [reply plugin/compile-scss.js:147]: I'm moving it to a separate PR

File: plugin/compile-scss.js (hunk 8 above)
Comment (Firfi) [reply plugin/compile-scss.js:336]: Unrelated indeed. Thanks!

File: plugin/compile-scss.js (hunk 7 above)
Comment (Firfi) [reply plugin/compile-scss.js:326]: I'm moving it into a separate PR

File: plugin/compile-scss.js (hunk 3 above)
Comment (Firfi) [reply plugin/compile-scss.js:143]: > Is this something required by dart-sass or is it an additional feature you like to include in meteor-scss package?

Rechecked with my project, doesn't seem to be needed indeed.

File: plugin/compile-scss.js (hunk 5 above)
Comment (Firfi) [reply plugin/compile-scss.js:152]: Thanks.

File: plugin/compile-scss.js (hunk 6 above)
Comment (Firfi) [reply plugin/compile-scss.js:293]: Checked with my app, doesn't seem to be needed indeed.

**Comment by Firfi** [comment]
Thank you for feedback @znerol . A new PR coming soon.

**Comment by Firfi** [comment]
@znerol opened https://github.com/Meteor-Community-Packages/meteor-scss/pull/296
